src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1129,45 +1207,59 @@ private bool HasCollectionInitializerTypeInProgress(SyntaxNode syntax, TypeSymbo
             return false;
         }
 
-        internal MethodSymbol? GetAndValidateCollectionBuilderMethod(
+        internal ImmutableArray<MethodSymbol> GetAndValidateCollectionBuilderMethods(
             SyntaxNode syntax,
-            NamedTypeSymbol namedType,
-            BindingDiagnosticBag diagnostics,
-            out TypeSymbol? elementType)
+            NamedTypeSymbol targetType,
+            TypeSymbol? builderType,
+            string? methodName,
+            BindingDiagnosticBag diagnostics)
         {
-            MethodSymbol? collectionBuilderMethod;
-            bool result = namedType.HasCollectionBuilderAttribute(out TypeSymbol? builderType, out string? methodName);
-            Debug.Assert(result);
+            Debug.Assert(targetType.IsDefinition);
 
-            var targetTypeOriginalDefinition = namedType.OriginalDefinition;
-            result = TryGetCollectionIterationType(syntax, targetTypeOriginalDefinition, out TypeWithAnnotations elementTypeOriginalDefinition);
+            bool result = TryGetCollectionIterationType(syntax, targetType, out var elementType);
             Debug.Assert(result);
 
-            var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
-            collectionBuilderMethod = GetCollectionBuilderMethod(namedType, elementTypeOriginalDefinition.Type, builderType, methodName, ref useSiteInfo);
-            diagnostics.Add(syntax, useSiteInfo);
-            if (collectionBuilderMethod is null)
+            ImmutableArray<MethodSymbol> candidates = ImmutableArray<MethodSymbol>.Empty;
+
+            if (SourceNamedTypeSymbol.IsValidCollectionBuilderType(builderType) &&
+                !string.IsNullOrEmpty(methodName))
             {
-                diagnostics.Add(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, syntax, methodName ?? "", elementTypeOriginalDefinition, targetTypeOriginalDefinition);
-                elementType = null;
-                return null;
-            }
+                var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                candidates = GetCollectionBuilderMethods(targetType, elementType.Type, (NamedTypeSymbol)builderType, methodName, ref useSiteInfo);
+                diagnostics.Add(syntax, useSiteInfo);
 
-            ReportUseSite(collectionBuilderMethod, diagnostics, syntax.Location);
+                Debug.Assert(candidates.All(static (m, n) => m.Arity == n, targetType.AllTypeArgumentCount()));
 
-            var parameterType = (NamedTypeSymbol)collectionBuilderMethod.Parameters[0].Type;
-            Debug.Assert(parameterType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions));
+                ReportDiagnosticsIfObsolete(diagnostics, builderType, syntax, hasBaseReceiver: false);
+            }
 
-            elementType = parameterType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+            if (candidates.IsEmpty)
```

Comment:
spec: To be considered a proper builder type, should we require at least the default `Create` method to be found (single ROS parameter)? Do we have or want collection types that require a with-element to be used?

Reply:
This is an open question.